### ISSUE TITLE
Weird address issues on ATtiny2313-10PU

### ISSUE THREAD

**Issue by iostat** [eriksl/usitwislave#4]
EDIT: I may be doing something wrong with my code, however, I've found that when configuring my ATtiny2313V-10PU to be an I2C slave on 0x40, probing the I2C bus with my raspberry pi using i2cdetect sometimes gets hits on just 0x40, sometimes on 0x40 and 0x41, sometimes on 0x40, 0x41, AND 0x42, and sometimes even 0x40, 0x41, 0x42, and 0x43.

I've blanked out this issue in the meantime while I investigate even further, but I have a feeling that something fishy is going on here. Here's my original code. I can confirm that that the address select pins are giving me a get_i2c_address() of 0b01000000 (0x40) using bit_bang_spi to drive a shift-register based LED driver. i2c_slave.h is of course, usitwislave.h

``` c
#include <avr/io.h>

#include "spi.h"
#include "i2c_slave.h"

#define REGISTER_COUNT 32
volatile uint8_t registers[REGISTER_COUNT];

void init_gpio()
{
    // ------------- ADDRESS SELECT -------------
    // set PA0 as input (address select pins)
    DDRA &= ~(1 << 0);

    // Set PD2,3,4,5 as input (address selection pins)
    DDRD &= ~(1 << 2);
    DDRD &= ~(1 << 3);
    DDRD &= ~(1 << 4);
    DDRD &= ~(1 << 5);

    // Enable pullups for address selection pins (PA0, PD2,3,4,5).
    PORTA |= (1 << 0);
    PORTD |= (1 << 2);
    PORTD |= (1 << 3);
    PORTD |= (1 << 4);
    PORTD |= (1 << 5);

    // -------------- BIT-BANG SPI --------------
    // set PA1 as output (LED driver latch-enable)
    DDRA |= (1 << 1);

    // Set PD0,1 as output (LED driver, serial clock and data)
    DDRD |= (1 << 0);
    DDRD |= (1 << 1);

    // ------------- ROTARY ENCODER -------------
    // Set PB0,1,2 as input
    DDRB &= ~(1 << 0);
    DDRB &= ~(1 << 1);
    DDRB &= ~(1 << 2);

    // Enable pullups for rotary encoder pins
    PORTB |= (1 << 0);
    PORTB |= (1 << 1);
    PORTB |= (1 << 2);
}

inline unsigned char get_i2c_address()
{
    unsigned char base = 0b01000000;
    base |= ((PINA & (1 << 0)) << 4); // mask out PA0 and shift to 5th bit
    base |= ((PIND & (1 << 2)) << 1); // mask out PD2 and shift to 4th bit
    base |= ((PIND & (1 << 3)) >> 1); // mask out PD3 and shift to 3rd bit
    base |= ((PIND & (1 << 4)) >> 3); // mask out PD4 and shift to 2nd bit
    base |= ((PIND & (1 << 5)) >> 5); // mask out PD5 and shift to 1st bit

    return base;

}

void idle_callback(void)
{
    //bit_bang_spi(get_i2c_address());
}

void i2c_callback(uint8_t input_buffer_length,
            const uint8_t* input_buffer, uint8_t* output_buffer_length,
            uint8_t* output_buffer)
{
    if(input_buffer_length > 0) {
        *output_buffer_length = 1;

        if(input_buffer[0] < REGISTER_COUNT) {
            *output_buffer = registers[input_buffer[0]];
        } else {
            *output_buffer = 0xFF;
        }
    }
}

void main(void)
{
    init_gpio();

    for(int i = 0; i < REGISTER_COUNT; i++) {
        registers[i] = i;
    }

    usi_twi_slave(get_i2c_address(), 0, i2c_callback, idle_callback);

    while(1);
}
```

**Comment by akohlsmith** [comment]
I'm (really) late to the party but I have noticed on the tiny25 that if you want anything even remotely close to 400kHz I2C bus speed you've got to be running up near 64MHz. I was quite surprised that even at 8MHz (internal RC, no PLL, CLKDIV8 not set) 100kHz was iffy. As soon as I enabled the PLL things just started working properly.

**Comment by gismofx** [comment]
@akohlsmith Where are you enabling the PLL? Can you post the line of code? I had some issues as well, and I started running my ATTINY85 with a 16mhz external crystal. It improved, but on occasion I get some problems. I'd like to try.

**Comment by eriksl** [comment]
Yes confirmed. You need to run > 8 Mhz to make I2C reliable. It can be done quite easily with the PLL clock (which the larger models don't have...), without a crystal. I2C is not sensitive to timing, so it doesn't matter that the RC clock isn't accurate. It just needs to be FAST, at least with the USI-implementation. The problem is in the acknowledgement of the address after a start condition. It's very easy to take to long for that and the master will give up. Unfortunately the USI-implementation cannot acknowledge in hardware. There is another problem with USI and that's there is no interrupt for STOP conditions. So you need to rely on STOP conditions always coming right before a START condition. Or monitor the line all of the time, but that's not what we want.

**Comment by akohlsmith** [comment]
@gismofx it's done in the fuses. http://www.engbedded.com/fusecalc/ shows that CKSEL=0001 will select the PLL (on ATTINY25).

**Comment by akohlsmith** [comment]
@eriksl that seems odd; USI's supposed to be taking the hard part away from software. I'm almost positive that I could bit-bang an I2C slave if I was running at 64MHz. No matter, I've got it figured out now. Do you think it might be worth adding a note to that in the readme?

**Comment by eriksl** [comment]
Funny you mention that. Definitely more people have mentioned this fact. The USI isn't helping very much indeed, I also found out. If you're going to do a full-blown I2C slave, you'd better switch to an atmega with full I2C support. The USI implementation would be sufficient for an I2C master though.

On a side note, how did you manage to get it running at 64 Mhz??? I am sure you can get the clock that high using the PLL, but it isn't supposed to work (max 20 Mhz!).

**Comment by akohlsmith** [comment]
I'm not doing anything spectacular. The ATTINY25's PLL takes a source clock and multiplies it by 8. The default input clock is the 8MHz internal RC oscillator. Section 6.1.5 of the datasheet. I am guessing that the 20MHz limit is the clock pin input limit.

**Comment by eriksl** [comment]
Or the clock is limited internally! Are you sure it really runs at the full 64 Mhz? With limited voltage (e.g. 3.3V) it can't even run on 20 Mhz...
